### DISCUSSION TITLE
Google Colaboratory VMs no longer support versions of TensorFlow 1.x

### DISCUSSION THREAD

**Discussion by senracg** [Denis2054/Transformers-for-NLP-2nd-Edition#10] [Q&A]
Google Colaboratory VMs no longer support versions of TensorFlow 1.x making Appendix IV of Transformers for Natural Language Processing - 2nd Edition incapable of runtime completion (section 4 and 9 of the coding).  Is there an updated workaround that can be used with Tensorflow 2.x that can simulate tensorflow.contrib from TensorFlow 1.x; now spread across TensorFlow 2.x?

**Comment by Denis2054** [comment]
There are two solutions:

Solution 1: Running this notebook locally after installing Tensforflow 1. x on a local machine and setting up the directories as defined in the notebook.
or (and)

Solution 2: Running a new 2023 notebook in the BONUS directory of this repository : [GPT-2 & ChatGPT-the Origins.ipynb](https://github.com/Denis2054/Transformers-for-NLP-2nd-Edition/blob/main/Bonus/GPT_2_and_ChatGPT_the_Origins.ipynb)
This new notebook contains:

a guide to understanding the source code of GPT-2

a Hugging Face implementation of text generation with GPT-2

text generation with OpenAI davinci-instruct series(GPT-3) which OpenAI ChatGPT was derived from. You will be able to run an example of a dialogue and compare the output of the davinci-instruct-beta engine with the output of ChatGPT
